lib/mkl/wrappers.jl

```diff
@@ -87,39 +87,19 @@ for (fname, elty) in
             Bptrs = unsafe_batch(B)
             Cptrs = unsafe_batch(C)
             bsize = length(A)
-            m_buf = oneVector{Int,oneL0.SharedBuffer}(undef,bsize)
-            n_buf = oneVector{Int,oneL0.SharedBuffer}(undef,bsize)
-            k_buf = oneVector{Int,oneL0.SharedBuffer}(undef,bsize)
-            lda_buf = oneVector{Int,oneL0.SharedBuffer}(undef,bsize)
-            ldb_buf = oneVector{Int,oneL0.SharedBuffer}(undef,bsize)
-            ldc_buf = oneVector{Int,oneL0.SharedBuffer}(undef,bsize)
-            alpha_buf = oneVector{$elty,oneL0.SharedBuffer}(undef,bsize)
-            beta_buf = oneVector{$elty,oneL0.SharedBuffer}(undef,bsize)
-            groupsize_buf = oneVector{Int,oneL0.SharedBuffer}(undef,bsize)
-
-            fill!(m_buf,m)
-            fill!(n_buf,n)
-            fill!(k_buf,k)
-            fill!(alpha_buf,alpha)
-            fill!(lda_buf,lda)
-            fill!(ldb_buf,ldb)
-            fill!(ldc_buf,ldc)
-            fill!(beta_buf,beta)
-            fill!(groupsize_buf,1)
-
-            m_ptr = unsafe_wrap(Array,m_buf)
-            n_ptr = unsafe_wrap(Array,n_buf)
-            k_ptr = unsafe_wrap(Array,k_buf)
-            alpha_ptr = unsafe_wrap(Array,alpha_buf)
-            lda_ptr = unsafe_wrap(Array,lda_buf)
-            ldb_ptr = unsafe_wrap(Array,ldb_buf)
-            ldc_ptr = unsafe_wrap(Array,ldc_buf)
-            beta_ptr = unsafe_wrap(Array,beta_buf)
-            groupsize_ptr = unsafe_wrap(Array,groupsize_buf)
+            m_dev = Vector{Int}(fill(m,bsize))
```

Comment:
This allocates a CPU vector. Don't you need device memory by using the `oneVector` constructor?

Reply:
I'm very surprised this passes CI...

Reply:
@maleadt : Thanks for catching this. Yes, they have changed the prototype now that it can take host pointer parameters as well that's why CI is passing. Please refer to prototype below,
https://github.com/oneapi-src/oneMKL/blob/e8c523e41490cb3d0807b13cf00b899cc416df98/src/blas/backends/mkl_common/mkl_batch.cxx#L645

We will stay with the original design of using device pointer i.e., oneVector and we will update it.

Reply:
> Yes, they have changed the prototype now that it can take host pointer parameters as well that's why CI is passing.

In that case, why not just use a Julia array? Or are there lifetime issues?

Reply:
We would like to have explicit allocations instead of having oneMKL to manage buffer transfers. 

For example, in terms of memory transfers:

1. **oneMKL Device Array API:** 
     1.1 If a parameter array is onevector then there is no need of parameter array copy before oneMKL call.
     1.2 If a parameter array is cpu array then Julia developer need to copy to device array before oneMKL call.
2. **oneMKL host array based API:**
    2.1 If a parameter array is cpu array: although there is no need to do anything by Julia developer but oneMKL performs the host/device copy underneath.
    2.2 If a parameter array is device array: Julia developer need to copy onevector to cpuarray before  onemkl call and then oneMKL performs host/device copy underneath which is very inefficient.

Using oneMKL device array API is better or equal for both cases.

Reply:
> 2. oneMKL host array based API:
> 2.1 If a parameter array is cpu array: although there is no need to do anything by Julia developer but oneMKL performs the host/device copy underneath.

That seems much easier? And the inputs to these functions are scalars, so it's not like we can avoid the copy.

Reply:
oneMKL's batch primitives are defined differently from cuBLAS.  All the parameters are arrays such that the shapes of the matrices can be different, so do the betas and group_sizes.  We currently only expose the scalar interface on the Julia function to match cuBLAS, but we may expose the full array parameter interface latter on if needed.  The current wrapper implementation will work just fine for the array parameter cases and benefit 2.2.

Can we get this merged and see how things turn out?  It should be an easy switch to call the host parameter array version if necessary.

Reply:
I don't understand why we would want to maintain this complexity now while oneMKL can do it for us, but yeah I guess we can merge this.